Add circular mean to statistics integration

enzo2 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
Add new characteristic to the statistics integration - circular mean. This allows averaging of circular/angular sensor data, e.g. wind direction, where arithmetic mean would give the wrong result when the value crosses over 0/360°. 

This would be useful, for example, to those running a weather station where the wind direction needs to be smoothed out. It could also be used for averaging a time of day, if converted to degrees (× 15) and then back to hours.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- See attached example from my local testing showing how it diverges from arithmetic mean.
![mean_circular vs mean 2_104814](https://github.com/home-assistant/core/assets/542271/6627e9d5-7f79-4bdb-a2f9-e695a6c239e7)

- This is my first PR to Home Assistant, please forgive any errors. 
- For the test, I didn't want to mess around too much with the tests for this component by making exceptions just for this one characteristic. I left the unit as °C like the other characteristics, whereas in reality it should be degrees (°). The numeric values used for the test could ideally be more widespread, i.e. past 360, to prove that it works. I did check the math, though.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/28682

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

enzo2 committed:
Add circular mean

Add support for circular mean for sensors in units of degrees, e.g. direction data.

enzo2 committed:
Update test_sensor.py

enzo2 committed:
Update sensor.py

enzo2 committed:
Remove whitespace

enzo2 committed:
Revert to degC

ThomDietrich commented on tests/components/statistics/test_sensor.py:
```diff
@@ -920,6 +921,24 @@ async def test_state_characteristics(hass: HomeAssistant) -> None:
+                        )
+                    ),
+                    2,
+                )
```
Hey, could you replace the formula by the actual result, or some simpler representation of it? When the test case uses the exact same logic as the main function, it's a self fulfilling promise :)

ThomDietrich commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -6,6 +6,7 @@
 import contextlib
 from datetime import datetime, timedelta
 import logging
+import math
```
I am honestly surprised `math` was not needed so far :)

ThomDietrich commented:
Hey @enzo2,
thanks for your contribution and congrats on your first PR over here.

The use case is clear and the code cleanly integrated with the existing. Regarding the test case and your concerns: Yes it was a good decision to simply stay with the example under `test_state_characteristics()`, however, I agree that your function would benefit from an extended example. Could you please define a new test function? I suggest `test_state_characteristic_mean_circular()` behind the general one.

enzo2 committed:
Fix: shift atan2 output to positive degrees

enzo2 committed:
Add new dedicated test

enzo2 commented:
Thanks! Kept the first test and simplified to the actual correct result, and added a new dedicated test that generates a set of data and uses a different complex numbers method. (Possibly went too far with this, but it works.) With the addition of cmath and numpy, I ran the gen_requirements_all script but nothing changed.

ThomDietrich commented:
Looking good but why did you build such a convoluted testcase? 🙈🙊 I was expecting a simple testcase like the one you can find under e.g. `test_sampling_size_reduced`. Fill the buffer and check the result. Instead if `VALUES_NUMERIC` use your own array:

```python
    VALUES_ANGULAR = [350, 10]  # should result in mean_circular==0?
```

enzo2 commented:
Honestly, I have no experience to know what is too rigorous versus too simplistic. Getting the hang of it. Will update.

enzo2 committed:
Simplify test

ThomDietrich commented:
In my own words, on the spot, I would say test cases need to be focused at one specific function (hence why I didn't like the overly complex function) and it needs to test all the relevant edge cases (hence why I am testing for 0 degrees instead of 360 degrees). If you can think of more edge cases in which the function works special, might be worth checking as well.

ThomDietrich commented on tests/components/statistics/test_sensor.py:
```diff
@@ -1207,6 +1216,43 @@ async def test_state_characteristics(hass: HomeAssistant) -> None:
+                    "name": "test_sensor_mean_circular",
+                    "entity_id": "sensor.test_monitored",
+                    "state_characteristic": "mean_circular",
+                    "sampling_size": 6,
```
sampling size isn't needed for the test case

enzo2 replied:
Is there a way to avoid it? For me the test will fail due to 
`Invalid config for [sensor.statistics]: The sensor configuration must provide 'max_age' and/or 'sampling_size'.`

ThomDietrich replied:
Ooooh nevermind. I myself was the one who made the change that shows your error message :D Nevermind

enzo2 commented:
The other edge cases I can think of are:
1. Can arctan2 handle all values, namely 0,0? Yes, it is defined as 0 in that case since the underlying C function follows the IEEE floating point standards. I cannot produce an error in testing, e.g. [90, 270], [-90, 90], [0, 180]
2. Evenly spaced data around the circle: Not a bug, just that there is no average possible in this case. The function comes up with an answer.

ThomDietrich reviewed (approved):
All looks good. PR can be merged from my point of view. @emontnemery

frenck reviewed (approved):
Looks good! Thanks, @enzo2 👍 

../Frenck
